packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -1,160 +1,144 @@
+      } | {
+        lifecycle: typeof lifecycleEnum.transient | KeyedSingleton;
+      })
+      : {
+        lifecycle: typeof lifecycleEnum.transient | KeyedSingleton;
+      }
+  );
 
   export function getInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam = void
   >(
-    options: InjectionTokenInjectable<TInjectionToken>,
-  ): InjectionTokenInjectable<TInjectionToken>;
+    options: Injectable<Instance, InjectionInstance, InstantiationParam>,
+  ): typeof options;
 
-  export function getInjectable<TInstance, TInstantiationParameter>(
-    options: NormalInjectable<TInstance, TInstantiationParameter>,
-  ): NormalInjectable<TInstance, TInstantiationParameter>;
-
-  export function getInjectionToken<TInstance, TInstantiationParameter = void>({
+  export function getInjectionToken<
+    Instance,
+    InstantiationParam = void,
+  >({
     id: string,
-  }): InjectionToken<TInstance, TInstantiationParameter>;
+  }): InjectionToken<Instance, InstantiationParam>;
+
+  type AsyncReturnable<IsAsync extends boolean, Instance> = IsAsync extends true
+    ? Promise<Instance>
+    : Instance;
+
+  export interface Inject<IsAsync extends boolean> {
+    <Instance>(
+      key: Injectable<Instance, any, void> | InjectionToken<Instance, void>,
+    ): AsyncReturnable<IsAsync, Instance>;
+    <Instance, InstantiationParam>(
+      key: Injectable<Instance, any, InstantiationParam> | InjectionToken<Instance, InstantiationParam>,
+      param: InstantiationParam,
+    ): AsyncReturnable<IsAsync, Instance>;
+  }
 
   interface DiContainerForInjection<TReturnAsPromise extends boolean> {
-    inject<TInjectable extends NormalInjectable<unknown, unknown>>(
-      injectableKey: TInjectable,
-      ...instantiationParameter: TentativeTuple<
-        InferFromInjectable<TInjectable>[1]
-      >
-    ): TReturnAsPromise extends true
-      ? Promise<InferFromInjectable<TInjectable>[0]>
-      : InferFromInjectable<TInjectable>[0];
-
-    inject<TInjectionToken extends InjectionToken<unknown, unknown>>(
-      injectionToken: TInjectionToken,
-      ...instantiationParameter: TentativeTuple<
-        TInjectionToken['instantiationParameter']
-      >
-    ): TReturnAsPromise extends true
-      ? Promise<TInjectionToken['template']>
-      : TInjectionToken['template'];
-
-    injectMany<TInjectionToken extends InjectionToken<unknown, unknown>>(
-      injectionToken: TInjectionToken,
-      ...instantiationParameter: TentativeTuple<
-        TInjectionToken['instantiationParameter']
-      >
-    ): TReturnAsPromise extends true
-      ? Promise<TInjectionToken['template'][]>
-      : TInjectionToken['template'][];
+    inject: Inject<TReturnAsPromise>;
+
+    injectMany<Instance, InstantiationParam = void>(
+      injectionToken: InjectionToken<Instance, InstantiationParam>,
+      instantiationParameter: InstantiationParam[],
```

Comment:
Will this change for not inferring `instantiationParameter` from `injectionToken` require explicit type parameters?